reconcile/terraform_repo.py

```diff
@@ -0,0 +1,302 @@
+import logging
+from collections.abc import Callable
+from typing import (
+    Any,
+    Optional,
+)
+
+import yaml
+from pydantic import (
+    BaseModel,
+    ValidationError,
+)
+
+from reconcile import queries
+from reconcile.gql_definitions.terraform_repo.terraform_repo import (
+    TerraformRepoV1,
+    query,
+)
+from reconcile.utils import gql
+from reconcile.utils.defer import defer
+from reconcile.utils.differ import (
+    DiffResult,
+    diff_iterables,
+)
+from reconcile.utils.exceptions import ParameterError
+from reconcile.utils.gitlab_api import GitLabApi
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.state import (
+    State,
+    init_state,
+)
+
+
+class RepoSecret(BaseModel):
+    path: str
+    version: Optional[int]
+
+
+class RepoOutput(BaseModel):
+    """
+    Output of the QR terraform-repo integration and input to the executor
+    which removes some information that is unnecessary for the executor to parse
+    """
+
+    dry_run: bool
+    repository: str
+    name: str
+    ref: str
+    project_path: str
+    delete: bool
+    secret: RepoSecret
+
+
+class TerraformRepoIntegrationParams(PydanticRunParams):
+    output_dir: Optional[str]
+    validate_git: bool
+
+
+class TerraformRepoIntegration(
+    QontractReconcileIntegration[TerraformRepoIntegrationParams]
+):
+    def __init__(self, params: TerraformRepoIntegrationParams) -> None:
+        super().__init__(params)
+        self.qontract_integration = "terraform_repo"
+        self.qontract_integration_version = make_semver(0, 1, 0)
```

Comment:
Am I imagining that this is handled in different locations depending on the integration? Is this our current best practice location? 

Alternative and off topic, is anyone using these for anything? Or is this just sounds-good code cruft?

Reply:
The boilerplate setup for integrations isn't standardized and varies a lot depending on the integration. This new `QontractReconcileIntegration` class is being used in a few newer integrations and was suggested by Olivia in https://github.com/app-sre/qontract-reconcile/pull/3464#discussion_r1179208282